KhronosGroup/KTX-Software
# JNI improvements

Addresses https://github.com/KhronosGroup/KTX-Software/issues/880 

I know, the title is too generic. It would be better to clearly lay out a work plan and create a clean sequence of self-contained, _specific_ PRs. But there are just _too many (unrelated) things to do_. "You can't cross a chasm in two small jumps". The first pass has to be tackled somewhat holistically. The result may not be as 'good' as I'd like it to be, but I hope that it will be 'better' along many dimensions. If this is not considered to be a viable approach, then that's OK. Then I'll just create my own repo with JNI bindings for KTX, with blackjack and error handling.

---

The preliminary task list that I tried to create in the [linked issue](https://github.com/KhronosGroup/KTX-Software/issues/880 ) is now replicated here, and I'll try to update it and extend it as we move on:

- [x] Wrap up https://github.com/KhronosGroup/KTX-Software/pull/876 
- [x] Wrap up https://github.com/KhronosGroup/KTX-Software/pull/879 
- [x] Expose `CreateFromMemory`
- [x] Improve the use of constants
  - There should be documentation for these constants, and `stringFor` ... methods to convert them to strings
  - Updates:
    - <strike>Aligned the names of all constants with the native version</strike>
    - Aligned the names of constants with the other bindings (see https://github.com/KhronosGroup/KTX-Software/pull/886#issuecomment-2308404357 )
    - Added documentation and string functions for all constants, except for...
       - `KtxInternalformat`
       - `VkFormat` 
       
       (See https://github.com/KhronosGroup/KTX-Software/issues/887 )

    - NOTES: 
      - The `ktxTexture2.supercompressionScheme` does not appear in the native documentation
      - The constants for `KtxPackAstcEncoderMode` are not documented on the native side
- [x] Fix the `KtxTestLibraryLoader` to also work on Windows
- [x] The JNI class- and field initializations should only be done once. Right now, things like `env->GetFieldID(ktx_texture_class, "instance", "J");` are called in each function. The `jfieldID` should be obtained once initially.
- [x] (continuously) Error checks, error checks, error checks ... (no JVM crashes, period!)
  - Updates (I did a first pass for this, but will keep an eye on places where checks have to be added):
    -  Passing `null` to any method should throw a `NullPointerException`. 
    - Trying to use a texture after `destroy()` was called should throw an `IllegalStateException`. 
    - When a JNI function causes an error (like an out-of-memory error), then the native functions have to return **immediately**, because there is a pending exception. This is not done in all places yet...
    - Java methods that do _not_ return error codes (like the `create...` methods) now generally throw a `KtxException` when the implementation caused an error code that was not `KTX_SUCCESS`
- [x] (continuously) Comments, comments, comments...
  - Updates: 
    - I have added _basic_ JavaDocs, and enabled the creation of the JavaDocs in the POM.
- [x] Add bindings for `ktxErrorString`

---

Not (yet) supposed to be addressed here:

- Try to find a clean solution for https://github.com/KhronosGroup/KTX-Software/issues/624 

- There are a few cases where we might have to think about how to handle the absence of `unsigned` types in Java. For example, when calling `createInfo.setBaseWidth(-123); // INVALID!`, then this value is implicitly converted into a `ktx_uint32_t`.

- <strike>Update the documentation of `deflateZstd` and `deflateZLIB`</strike>
  EDIT: This was now done as part of https://github.com/KhronosGroup/KTX-Software/pull/879, but only for the native side and for the Java bindings (other bindings are still missing the doc update or the functions)
  - As pointed out in [a comment](https://github.com/KhronosGroup/KTX-Software/pull/879#discussion_r1537201514), this documentation should include the information that 'the data pointer is updated'. This will go into the [native documentation](https://github.com/KhronosGroup/KTX-Software/blob/5414bd0ebfdfdf4ecb8c613a6edb21372303c2b9/lib/writer2.c#L762) first, but it will also have to be passed through to the [python bindings](https://github.com/KhronosGroup/KTX-Software/blob/7b08daaaa2f1271d0040d9307156c651f55ad58a/interface/python_binding/pyktx/ktx_texture2.py#L168), the Java binding, and maybe to the JS binding. **But...** the `deflateZLIB` function is still missing in the Python binding (and the `deflate*` functions are just about to be added to the Java binding), so ... this may have to be tracked and addressed separately

- Improve thoroughness of `deflate*` tests by comparing the deflated values
  - As pointed out in [another comment](https://github.com/KhronosGroup/KTX-Software/pull/879#discussion_r1537209791), the unit test for the `deflate*` functions should compare the result of _inflating_ the data again to the original input. The `inflate*` methods are not yet exposed (anywhere), so this may have to be deferred for now

- Add test for input swizzling in ASTC
  - The [PR for fixing the swizzling in JNI](https://github.com/KhronosGroup/KTX-Software/pull/876) includes a test for the swizzling for Basis. It does _not_ include one for ASTC. These are different code paths, and should both be tested. But right now, I'm not sure _what_ could be checked for the ASTC result (see [this PR comment](https://github.com/KhronosGroup/KTX-Software/pull/876#discussion_r1539044357) for context).

- Consider consistent formatting. I know, it's a detail, but just auto-formatting everything with the Eclipse default could already be an improvement. I do have strong preferences for my own projects, but here, the main point is that the formatting should be consistent...
 
- The [ktxHash...](https://github.khronos.org/KTX-Software/libktx/classktxHashList.html) functionalities are not yet mapped through from the native side.
  - Consider following the new JS binding here. It adds `findKeyValue`, `addKVPair` and `deleteKVPair` methods to the ktxTexture2 object rather than exposing `ktxHashList`.
